### PR TITLE
Allow to downgrade local volumes from > 1.7 to 1.6.

### LINKED ISSUE #13695: Downgrade from 1.7 to 1.6 doesn't work
Before going into exactly what is failing I have to first ask the question whether this is supported.  If a user upgrades to 1.7 and then downgrade to 1.6, should things still work?  I would say yes they should as that has always seemed to work in the past.  With 1.7 I'm getting errors now.  Basically after reverting back to 1.6 I get errors like:

```time="2015-06-03T07:43:11Z" level=fatal msg="reflect: call of reflect.Value.Type on zero Value" ```

### PR DESCRIPTION
**Pull request by calavera**
dockercompat

See for details #13695

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
ping @ibuildthecloud @cpuguy83 PR for the downgrade issue

**Comment by crosbymichael** [comment]
Downgrades have never worked

**Comment by icecrime** [comment]
@crosbymichael Is that to say that we shouldn't even consider this PR?

**Comment by duglin** [comment]
I wouldn't mind a little bit of investigation to see why things are breaking - just to make sure its not exposing a hidden bug.

**Comment by icecrime** [comment]
[Agreed on IRC](https://botbot.me/freenode/docker-dev/2015-06-03/?msg=40911351&page=5) that we should close this PR.

**Comment by icecrime** [comment]
Reopening in light of the recent conversations on #13695. Does this fully resolve the issue?

**Comment by calavera** [comment]
There is some data migration handling that I can fix here.

**Comment by calavera** [comment]
This changes the local storage layout a little bit to allow Downgrade Docker 1.7 to Docker 1.6, **only** if a container was created with Docker < 1.7.

History:

Pre 1.7m, volume information was stored in two separated directories:

1. The volume configuration was in `ROOT/volumes/VOLUME_ID/config.json`.
2. The data stored in the volumes was in `ROOT/vfs/dir/VOLUME_ID`.

Post 1.7, volume information is stored in one place, `ROOT/volumes/VOLUME_ID/_data`.

The docker daemon takes care of migrating volumes from 1.6 to 1.7 by creating a symlink from `ROOT/vfs/dir/VOLUME_ID` to `ROOT/volumes/VOLUME_ID/_data`. The file `config.json` is preserved, so you can boot the container with Docker 1.6 and all the information is in the same place that it was before booting Docker 1.7.

Unfortunately, the two previous RCs for 1.7 didn't follow that migration, so there is a special migration for volumes created with Docker 1.7 RC1 and Docker 1.7 RC2. This is a very small corner case, I don't expect anyone running Docker 1.7 RC in production, except us, ouch.

**Comment by thaJeztah** [comment]
Thanks @calavera! Is an additional test needed to check if `docker rm -v` properly cleans up all locations and the symlinks?

**Comment by calavera** [comment]
good catch @thaJeztah, fixed.

**Comment by thaJeztah** [comment]
Thanks! :heart:

**Comment by icecrime** [comment]
I'm not sure we should bother with previous RC migration paths. Nevertheless, LGTM: I'm not 100% confident because the change is quite significant, but reading several times through it it seems good (and thanks for the code comments btw).

File: daemon/volumes.go [hunk 1]
```diff
@@ -241,53 +261,77 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
...
-	jsonPath, err := container.jsonPath()
-	if err != nil {
-		return err
-	}
-	f, err := os.Open(jsonPath)
-	if err != nil {
-		if os.IsNotExist(err) {
-			return nil
+	// Inspect old structures only when we're upgrading from old versions
+	// to > 1.7 and the MountPoints has not been populated with volumes data.
```
Comment (tiborvass) [inline daemon/volumes.go]: @calavera `>= 1.7`

File: daemon/volumes.go [hunk 2]
```diff
@@ -241,53 +261,77 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
...
+				// We should not find an error here coming
+				// from the old configuration, but who knows.
+				if err != nil {
+					return err
+				}
+				container.addBindMountPoint(id, source, destination, rw)
+			}
+		}
+	} else if len(container.MountPoints) > 0 {
+		// Volumes created post 1.6. We verify integrity in case of data created
```
Comment (tiborvass) [inline daemon/volumes.go]: > `post 1.6`

do you mean `version >= 1.7` ?

File: daemon/volumes.go [hunk 3]
```diff
@@ -241,53 +261,77 @@ func (daemon *Daemon) registerMountPoints(container *Container, hostConfig *runc
...
+		// with one of the 1.7 RC versions that put the information in
```
Comment (tiborvass) [inline daemon/volumes.go]: how is this specific to the RC versions and not to 1.7 in general?

Comment (calavera) [inline daemon/volumes.go]: When this PR is merged, this won't be a case anymore, the volume will be mounted in the right directory. See https://github.com/docker/docker/pull/13699#issuecomment-109424147

File: daemon/volumes.go (hunk 2 above)
Comment (calavera) [inline daemon/volumes.go]: `post 1.6` == `version >= 1.7` doesn't it? But I can change the wording

**Comment by calavera** [comment]
@tiborvass I think the comments are more clear now, let me know what you think.

**Comment by icecrime** [comment]
I was concerned about using prefix test on paths at first, but considering that `filepath.EvalSymLink` does a `filepath.Clean` before returning I think we're safe.

Still LGTM.

**Comment by tiborvass** [comment]
LGTM

**Comment by jfrazelle** [comment]
cherry-picked

**Pull request merged**